tools/distpartitioning/data_shuffle.py

```diff
@@ -199,83 +203,77 @@ def exchange_node_features(rank, world_size, node_data, node_features, ntypes_ma
-                feat_per_proc = math.ceil(global_feature_count / world_size)
-                global_type_nid_start = feat_per_proc * rank
-                global_type_nid_end = global_type_nid_start
-                if((global_type_nid_start + feat_per_proc) > global_feature_count):
-                    global_type_nid_end += (ntype_id_count[str(ntype_id)] - global_type_nid_start)
-                    type_nid = np.arange(0, (ntype_id_count[str(ntype_id)] - global_type_nid_start))
-                else: 
-                    global_type_nid_end += feat_per_proc 
-                    type_nid = np.arange(0, feat_per_proc)
-
-                #now map the global_ntype_id to global_nid 
-                global_nid_offset = ntypes_nid_map[ntype_name][0]
-                global_nid_start = global_type_nid_start + global_nid_offset
-                global_nid_end = global_type_nid_end + global_nid_offset
-
-                #assert (global_nid_end - global_nid_start) == feature_count
-                global_nids = np.arange(global_nid_start, global_nid_end, dtype=np.int64)
-
-                #determine node feature ownership 
-                part_ids_slice = node_part_ids[global_nid_start:global_nid_end]
-                idx = (part_ids_slice == part_id)
-                out_global_nid = global_nids[idx == 1]
-                out_type_nid = type_nid[idx == 1]
-                out_features = node_features[ntype_name+'/feat'][out_type_nid]
-                send_node_features[ntype_name+'/feat'] = out_features
-                send_global_nids[ntype_name+'/feat'] = out_global_nid
-
-        node_features_rank_lst.append(send_node_features)
-        global_nid_rank_lst.append(send_global_nids)
+    own_node_features = {}
+    own_global_nids = {}
+    for ntype_name, ntype_info in node_feature_tids.items():
+        for feat_info in ntype_info:
+            #determine the owner process for these node features. 
+            node_feats_per_rank = []
+            global_nid_per_rank = []
+            feat_name = feat_info[0]
+            feat_key = ntype_name+'/'+feat_name
+
+            #compute the global_nid range for this node features
+            type_nid_start = int(feat_info[1])
+            type_nid_end = int(feat_info[2])
+            begin_global_nid = ntype_gnid_map[ntype_name][0]
+            gnid_start = begin_global_nid + type_nid_start
+            gnid_end = begin_global_nid + type_nid_end
+
+            #type_nids for this feature subset on the current rank
+            gnids_feat = np.arange(gnid_start, gnid_end)
+            tnids_feat = np.arange(type_nid_start, type_nid_end)
+            local_idx = np.arange(0, type_nid_end - type_nid_start)
+
+            #check if node features exist for this ntype_name + feat_name
+            #this check should always pass, because node_feature_tids are built
+            #by reading the input metadata json file for existing node features.
+            if(feat_key in node_features):
```

Comment:
should we use assert instead of using if? when it fails?

Reply:
Agreed. changed the if-statement to assert, per suggestion.